aten/src/ATen/core/function_schema.h

```diff
@@ -156,18 +156,27 @@ struct FunctionSchema {
     checkSchema();
   }
 
-  // check whether this schema is backward compatible with the old one.
-  // the following conditions are considered as this schema is backward
-  // compatible with old:
-  //   1) two schemas are equal
-  //   2) this schema has the same or more positional args than old,
-  //      and any positional arg in this schema is backward compatible
-  //      with the corresponding one in old schema, which could be an arg
-  //      or a kwarg, if it has, or it must provide a default value
-  //   3) this schema has the same or more kwargs than old, and all the kwargs
-  //      in old schema can find the corresponding kwarg in this schema which
-  //      is backward compatible with the old kwarg, and the extra kwargs in
-  //      this schema must provide default values.
+  // Checks whether this schema is backward compatible with the old one.
+  // The following conditions must be true:
+  // [Function structure] The new schema's name, overload-name, varargs, and
+  //      return arity are the same.
+  // [Output Narrowing] The new schema's output type must be the same class
+  //      or inherit from the old schema's output type
+  // [Argument count] The new schema must have at least as many arguments as
+  //      the old schema (considering the list of positional and kwargs).
+  // [Arg Compatibility] Every argument in the old schema has a corresponding
```

Comment:
I like your organization. Should "Arg Compatibility" also cover the movement of args from positional to keyword-only? Here's the cases I'm thinking of

```
foo(a, b, *, c) -> foo(a, b, c)
# this seems OK?

foo(a, b, *, c) -> foo(a, *, b, c)
# this seems not OK
```

Reply:
Yes, you're right the second case "kwarg to positional being BC" is covered in the tests:
# BC: A new schema where old kwargs becomes positional: foo_new(a, b, *, c) -> foo_old(a, *, b, c)
Along with:
# BC: A new schema where all old kwargs become positional: foo_new(a, b, c) -> foo_old(a, *, b, c)

And indeed the opposite ("positional to kwarg not being BC") is not covered neither in the docs nor tests.

I just added a couples of test to cover what you mention:
# No-BC: A new schema where an old positional argument becomes kwarg: foo_new(a, *, b, c) -> foo_old(a, b, *, c)
# No-BC: A new schema where many old positional arguments become kwarg: foo_new(a, *, b, c) -> foo_old(a, b, c)

Also updated the docs to reflect this asymmetry.

Thanks!